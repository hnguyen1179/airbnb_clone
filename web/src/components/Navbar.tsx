import React, { useContext } from "react";
import useLogout from "../hooks/useLogout";
import AppContext from "../context/app-context";
import Loading from "./Loading";

function Navbar() {
  const { user } = useContext(AppContext);
  const handleLogout = useLogout();
	return (
    <div>
      <Loading />
      {user ? <h1>hello, {user.firstName}</h1> : <a href="/">sign in</a>}
      {user ? <button onClick={handleLogout}>logout</button> : ""}
		</div>
	);
}

export default Navbar;
