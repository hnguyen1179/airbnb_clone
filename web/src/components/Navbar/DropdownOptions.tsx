import React from "react";

interface Props {
	handleOpen: (e: React.SyntheticEvent<EventTarget>) => void;
}

const DropdownOptions = ({ handleOpen }: Props) => {
	return (
		<>
			<button
				className="Navbar__right__profile__dropdown__link"
				onClick={handleOpen}
			>
				<div>Log in</div>
			</button>
			<button
				className="Navbar__right__profile__dropdown__link"
				onClick={handleOpen}
			>
				<div>Sign up</div>
			</button>
		</>
	);
};

export default DropdownOptions;
