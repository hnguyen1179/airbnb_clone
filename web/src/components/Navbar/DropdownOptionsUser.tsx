import { NavLink } from "react-router-dom";

interface Props {
	user: any;
}

const DropdownOptionsUser = ({ user }: Props) => {
	return (
		<>
			<button className="Navbar__right__profile__dropdown__link">
				<NavLink to={`/user/${user.id}`}>
					<div>
						<span>Profile</span>
					</div>
				</NavLink>
			</button>
			<button className="Navbar__right__profile__dropdown__link">
				<NavLink to={`/trips/${user.id}`}>
					<div>
						<span>Trips</span>
					</div>
				</NavLink>
			</button>
		</>
	);
};

export default DropdownOptionsUser;
