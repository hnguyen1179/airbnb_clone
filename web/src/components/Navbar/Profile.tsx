import React, { useContext } from "react";
import { ReactComponent as ProfileSvg } from "../../assets/icons/dark-profile.svg";
import { ReactComponent as HamburgerSvg } from "../../assets/icons/hamburger.svg";
import { AppContext } from "../../context/AppContext";
import ProfileOptions from "./ProfileOptions";
import UserProfileOptions from "./UserProfileOptions";
import useLogout from "../../hooks/useLogout";

import { AdvancedImage, placeholder } from "@cloudinary/react";
import { CloudinaryImage } from "@cloudinary/base";

interface Props {
	profile: boolean;
	setProfile: (x: boolean) => void;
	handleOpen: (e: React.SyntheticEvent<EventTarget>) => void;
}

const Profile = ({ profile, setProfile, handleOpen }: Props) => {
	const logout = useLogout();
	const { cloudinary, user } = useContext(AppContext);

	const handleClickProfile = (e: React.SyntheticEvent<EventTarget>) => {
		e.stopPropagation();
		setProfile(!profile);
	};

	const handleLogOut = async () => {
		await logout();
		setProfile(false);
	};

	const profileActive = profile ? "active" : "";

	return (
		<div className="Navbar__right__profile">
			<button
				className="Navbar__right__profile__button"
				onClick={handleClickProfile}
			>
				<div className="Navbar__right__profile__button__icon Navbar__right__profile__button__icon--hamburger">
					<HamburgerSvg />
				</div>
				<div className="Navbar__right__profile__button__icon Navbar__right__profile__button__icon--profile">
					{user ? (
						<AdvancedImage
							style={{ width: "50px" }}
							cldImg={cloudinary.image(`user_avatars/${user.id}`)}
						/>
					) : (
						<ProfileSvg />
					)}
				</div>
			</button>

			<div
				className={`Navbar__right__profile__dropdown ${profileActive}`}
				onClick={(e) => e.stopPropagation()}
			>
				{/* Display these conditionally on the presence of a 'user' */}
				{user ? (
					<UserProfileOptions user={user} />
				) : (
					<ProfileOptions handleOpen={handleOpen} />
				)}

				{/* These are just placeholders */}
				<div className="divider"></div>
				<button className="Navbar__right__profile__dropdown__link Navbar__right__profile__dropdown__link--inactive">
					<span>Host your home</span>
				</button>
				<button className="Navbar__right__profile__dropdown__link Navbar__right__profile__dropdown__link--inactive">
					<span>Host an experience</span>
				</button>

				{user && <div className="divider"></div>}

				<button className="Navbar__right__profile__dropdown__link Navbar__right__profile__dropdown__link--inactive">
					<span>Help</span>
				</button>

				{user && (
					<button
						className="Navbar__right__profile__dropdown__link"
						onClick={handleLogOut}
					>
						<span>Log out</span>
					</button>
				)}
			</div>
		</div>
	);
};

export default Profile;
