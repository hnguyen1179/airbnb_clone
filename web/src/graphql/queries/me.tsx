import { gql } from '@apollo/client'

const meQuery = gql`
	query meQuery {
		me {
			id
			firstName
			lastName
		}
	}
`;

export { meQuery };