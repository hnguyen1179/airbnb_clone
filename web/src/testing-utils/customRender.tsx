import { render } from "@testing-library/react";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { ModalProvider } from "../context/ModalContext";

export const customRender = (
	node: JSX.Element | null,
	mocks?: MockedResponse[]
) => {
	return {
		...render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ModalProvider>
          {node}
        </ModalProvider>
			</MockedProvider>
		),
	};
};
