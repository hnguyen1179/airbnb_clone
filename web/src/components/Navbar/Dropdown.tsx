import { SyntheticEvent, useMemo } from "react";
import { ReactComponent as ProfileSvg } from "../../assets/icons/dark-profile.svg";
import { ReactComponent as HamburgerSvg } from "../../assets/icons/hamburger.svg";
import { useAppState } from "../../context/AppContext";
import DropdownOptions from "./DropdownOptions";
import DropdownOptionsUser from "./DropdownOptionsUser";
import useLogout from "../../hooks/useLogout";

import { AdvancedImage, placeholder } from "@cloudinary/react";
import { useHistory, useLocation } from "react-router";

interface Props {
	dropdown: boolean;
	setDropdown: (x: boolean) => void;
	handleOpen: (e: SyntheticEvent<EventTarget>) => void;
}

const Dropdown = ({ dropdown, setDropdown, handleOpen }: Props) => {
	const { cloudinary, user } = useAppState();
	const location = useLocation();
	const history = useHistory();
	const logout = useLogout();

	const handleClickDropdown = (e: SyntheticEvent<EventTarget>) => {
		e.stopPropagation();
		setDropdown(!dropdown);
	};

	const handleLogOut = async () => {
		setDropdown(false);

		if (user && location.pathname.includes(user?.id)) {
			await logout();
			history.push("/");
		} else {
			await logout();
		}
	};

	const renderAvatar = useMemo(() => {
		return user ? (
			<AdvancedImage
				style={{ width: "50px" }}
				cldImg={cloudinary.image(`user_avatars/${user.id}`)}
				plugins={[placeholder("predominant-color")]}
			/>
		) : (
			<ProfileSvg />
		);
	}, [cloudinary, user]);

	const dropdownActive = dropdown ? "active" : "";

	return (
		<div className="Navbar__right__profile">
			<button
				aria-label="Profile Dropdown Button"
				className="Navbar__right__profile__button"
				onClick={handleClickDropdown}
			>
				<div className="Navbar__right__profile__button__icon Navbar__right__profile__button__icon--hamburger">
					<HamburgerSvg />
				</div>
				<div className="Navbar__right__profile__button__icon Navbar__right__profile__button__icon--profile">
					{renderAvatar}
				</div>
			</button>

			<div
				className={`Navbar__right__profile__dropdown ${dropdownActive}`}
				onClick={(e) => e.stopPropagation()}
			>
				{/* Display these conditionally on the presence of a 'user' */}
				{user ? (
					<DropdownOptionsUser user={user} />
				) : (
					<DropdownOptions handleOpen={handleOpen} />
				)}

				{/* These are just placeholders */}
				<div className="navbar-divider"></div>
				<button className="Navbar__right__profile__dropdown__link Navbar__right__profile__dropdown__link--inactive">
					<span>Host your home</span>
				</button>
				<button className="Navbar__right__profile__dropdown__link Navbar__right__profile__dropdown__link--inactive">
					<span>Host an experience</span>
				</button>

				{user && <div className="navbar-divider"></div>}

				<button className="Navbar__right__profile__dropdown__link Navbar__right__profile__dropdown__link--inactive">
					<span>Help</span>
				</button>

				{user && (
					<button
						className="Navbar__right__profile__dropdown__link"
						onClick={handleLogOut}
					>
						<span>Log out</span>
					</button>
				)}
			</div>
		</div>
	);
};

export default Dropdown;
