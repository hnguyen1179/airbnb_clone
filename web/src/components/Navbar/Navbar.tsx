import React, { useEffect, useState, useRef } from "react";
import Modal from "@material-ui/core/Modal";
import { TransitionGroup, CSSTransition } from "react-transition-group";

import { useModal } from "../../context/ModalContext";

import Entry from "../Entry/Entry";
import Fade from "../Fade";
import { ReactComponent as LogoSvg } from "../../assets/icons/logo.svg";
import { ReactComponent as LogoNameSvg } from "../../assets/icons/logo-with-name.svg";
import { ReactComponent as GlobeSvg } from "../../assets/icons/globe.svg";

import { ReactComponent as SearchSvg } from "../../assets/icons/thick-search.svg";
import Profile from "./Profile";
import SearchForm from "./SearchForm";

interface Props {
	isTop?: boolean; // Denotes whether or not Navbar is at the top of page or not; used in Landing
	notLanding?: boolean; // Denotes whether or not Navbar is used on LandingPage or not
	disableEntry?: boolean; // Prevent login and signup modal from appearing
}

// This Navbar is for desktop view; contains a search
function Navbar({ isTop, notLanding = false, disableEntry = false }: Props) {
	const componentRef = useRef<HTMLDivElement>(null);
	const buttonRef = useRef<HTMLButtonElement>(null);
	const modalRef = useRef<HTMLDivElement>(null);
	const [profile, setProfile] = useState(false);
	const [search, setSearch] = useState(false);

	// Good react pattern in useModal; incorporate hooks within Context APIs that check for undefined
	const { open, setOpen, setEntry } = useModal();

	// Opens modal; closes profile and search components
	const handleOpen = (e: React.SyntheticEvent<EventTarget>) => {
		e.stopPropagation();
		if (disableEntry) {
			setProfile(false);
			return;
		}

		document.body.style.overflow = "hidden";
		setEntry("unverified");
		setSearch(false);
		setProfile(false);
		setOpen(true);
	};

	// Closes modal
	const handleClose = () => {
		document.body.style.overflow = "initial";
		setOpen(false);
	};

	// Opens search component
	const handleClickSearch = (e: React.SyntheticEvent<EventTarget>) => {
		e.stopPropagation();
		setProfile(false);
		setSearch(true);
	};

	// Closes profile and search component on document body click
	const handleBodyClick = () => {
		const closeNavigation = new Event("closeNavigation");
		window.dispatchEvent(closeNavigation);
	};

	// Clicking on the body while search or profile is opened will close either
	const handleCloseNavigation = () => {
		setSearch(false);
		setProfile(false);
	};

	// Activates the Search component if at the very t op of screen
	useEffect(() => {
		if (notLanding) {
			return;
		}

		console.log(" in here ");
		if (isTop) {
			setSearch(true);
		} else {
			setSearch(false);
		}
	}, [isTop]);

	// Creates event listeners to click out of profile and search components
	useEffect(() => {
		window.addEventListener("closeNavigation", handleCloseNavigation);
		document.body.addEventListener("click", handleBodyClick);

		return () => {
			window.removeEventListener(
				"closeNavigation",
				handleCloseNavigation
			);
			document.body.removeEventListener("click", handleBodyClick);
			handleClose();
		};
	}, []);

	const active = search ? "active" : "";
	const transparent = isTop ? "transparent" : "";

	return (
		<nav className={`Navbar ${active} ${transparent}`}>
			<div
				className="Navbar__container"
				onClick={(e) => e.stopPropagation()}
			>
				<div
					className="Navbar__left"
					onClick={(e) => {
						setProfile(false);
						e.stopPropagation();
					}}
				>
					<div className="Navbar__left__logo">
						<a href="/">
							<LogoNameSvg className="logo-name" />
							<LogoSvg className="logo" />
						</a>
					</div>
				</div>
				{!notLanding && (
					<div className="Navbar__search">
						<CSSTransition
							in={search || isTop}
							timeout={150}
							unmountOnExit
							classNames="component"
							nodeRef={componentRef}
						>
							<div
								className="Navbar__search__component"
								ref={componentRef}
							>
								<div className="Navbar__search__component__categories">
									<span>Places to stay</span>
									<span>Experiences</span>
									<span>Online Experiences</span>
								</div>
								<SearchForm />
							</div>
						</CSSTransition>
						<CSSTransition
							in={!search && !isTop}
							timeout={150}
							unmountOnExit
							classNames="button"
							nodeRef={buttonRef}
						>
							<button
								className="Navbar__search__button"
								ref={buttonRef}
								onClick={handleClickSearch}
							>
								<div>Start your search</div>
								<SearchSvg />
							</button>
						</CSSTransition>
					</div>
				)}

				<div className="Navbar__right">
					<div className="Navbar__right__host">
						<span>Become a host</span>
					</div>
					<div className="Navbar__right__globe">
						<button>
							<GlobeSvg />
						</button>
					</div>
					<Profile
						profile={profile}
						setProfile={setProfile}
						handleOpen={handleOpen}
					/>
				</div>

				{/* Modal for the entry form */}
				<Modal open={open} onClose={handleClose} disableScrollLock>
					<TransitionGroup>
						<Fade in={open} ref={modalRef}>
							<div className="Modal-container">
								<Entry />
							</div>
						</Fade>
					</TransitionGroup>
				</Modal>
			</div>
			<div className="background"></div>
		</nav>
	);
}

export default Navbar;
