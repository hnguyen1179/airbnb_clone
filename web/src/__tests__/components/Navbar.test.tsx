import { fireEvent, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import Navbar from "../../components/Navbar";
import { customRender } from "../../testing-utils/customRender";
import { meQuery } from "../../graphql/queries/me";

const meUserPresentMock = {
	request: {
		query: meQuery,
	},
	result: {
		data: {
			user: {
				firstName: "User",
				lastName: "Demo",
			},
			token: "Some Token",
		},
	},
};

const meUserNullMock = {
	request: {
		query: meQuery,
	},
	result: {
		data: {
			user: null,
			token: null,
		},
	},
};

afterEach(cleanup);
const waitForData = () => new Promise((res) => setTimeout(res, 0));

describe("<Navbar />", () => {
	it("renders the <Navbar /> with login and signup buttons", async () => {
		const { container, getByText } = customRender(<Navbar />, [
			meUserNullMock,
		]);

		await waitForData();

		expect(getByText(/log in/i)).toBeInTheDocument();
		expect(getByText(/sign up/i)).toBeInTheDocument();
		expect(container.firstChild).toMatchSnapshot();
	});

  // Navbar relies on an external AppState which queries 
	// it("renders the user's new if logged in", async () => {
	// 	const { getByTestId } = customRender(<Navbar />, [meUserPresentMock]);

	// 	await waitForData();

	// 	expect(getByTestId("greeting")).toBe("Hello, User");
	// });

	it("renders the correct modal depending on button", async () => {
		const { getByText } = customRender(<Navbar />, [meUserNullMock]);

		await waitForData();

		fireEvent.click(getByText(/log in/i));
		expect(getByText(/welcome/i)).toBeInTheDocument();

		fireEvent.click(getByText(/sign up/i));
		expect(getByText(/sign up/i)).toBeInTheDocument();
	});
});
