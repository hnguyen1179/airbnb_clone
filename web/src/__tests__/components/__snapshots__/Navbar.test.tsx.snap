// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar /> renders the <Navbar /> with login and signup buttons 1`] = `
<div
  class="Navbar-container"
>
  <div
    class="Loading-container"
  >
    <div
      class="Loading-container__dot"
    />
    <div
      class="Loading-container__dot"
    />
    <div
      class="Loading-container__dot"
    />
    <div
      class="Loading-container__dot"
    />
  </div>
  <div>
    <button
      value="signup"
    >
      sign up
    </button>
    <button
      value="unverified"
    >
      log in
    </button>
  </div>
  
</div>
`;
